File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  while(true){
    string str;
    getline(cin, str);

    if(str == ".") break;
    stack<char> s;

    bool flag = false;
    for(char c:str){
      if(c == '(' || c == '[') s.push(c);
      else if(c == ')'){
        if(s.empty() || s.top() != '('){
          flag = true;
          break;
        }
        s.pop();
      } else if(c == ']'){
        if(s.empty() || s.top() != '['){
          flag = true;
          break;
        }
        s.pop();
      }
    }

    if(!s.empty()) flag = true;
    flag ? cout << "no\n" : cout << "yes\n";
  }

}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        while(true){
            String str = br.readLine();

            if(str.equals(".")) break;
            Stack<Character> s = new Stack<>();
            boolean flag = false;
            for(int i=0; i<str.length(); i++){
                if(str.charAt(i) == '(' || str.charAt(i) == '[') s.push(str.charAt(i));
                else if(str.charAt(i) == ')'){
                    if(s.empty() || s.peek() != '('){
                        flag = true;
                        break;
                    }
                    s.pop();
                } else if(str.charAt(i) == ']'){
                    if(s.empty() || s.peek() != '['){
                        flag = true;
                        break;
                    }
                    s.pop();
                }
            }
            
            if(!s.empty()) flag = true;
            
            if(flag) System.out.println("no");
            else System.out.println("yes");
        }
    }

}
